src/main/java/bio/terra/dao/FileDao.java

```diff
@@ -0,0 +1,324 @@
+        }
+        createList.set(lastIndex, fileToCreate);
+
+        UUID id = null;  // set null to avoid warning, but due to the check above, we never return null
+        for (FSObject fsObject : createList) {
+            id = createObject(fsObject);
+        }
+        return id;
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public UUID createFileComplete(FSObject fsObject) {
+        FSObject fileNotPresent = retrieveFileByIdNoThrow(fsObject.getObjectId());
+        if (fileNotPresent.getObjectType() != FSObject.FSObjectType.FILE_NOT_PRESENT) {
+            throw new CorruptMetadataException("Unexpected file system object type");
+        }
+        if (!StringUtils.equals(fileNotPresent.getCreatingFlightId(),
+            fsObject.getCreatingFlightId())) {
+            throw new CorruptMetadataException("Unexpected flight id on file");
+        }
+
+        String sql = "UPDATE fs_object" +
+            " SET object_type = :object_type, gspath = :gspath," +
+            " checksum_crc32c = :checksum_crc32c, checksum_md5 = :checksum_md5, size = :size" +
+            " WHERE object_id = :object_id";
+        MapSqlParameterSource params = new MapSqlParameterSource()
+            .addValue("object_id", fsObject.getObjectId())
+            .addValue("object_type", FSObject.FSObjectType.FILE.toLetter())
+            .addValue("gspath", fsObject.getGspath())
+            .addValue("checksum_crc32c", fsObject.getChecksumCrc32c())
+            .addValue("checksum_md5", fsObject.getChecksumMd5())
+            .addValue("size", fsObject.getSize());
+        jdbcTemplate.update(sql, params);
+        return fsObject.getObjectId();
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public void createFileCompleteUndo(FSObject fsObject) {
+        // Reset to mark the file not preset
+        FSObject file = retrieveFileByIdNoThrow(fsObject.getObjectId());
+        if (!StringUtils.equals(file.getCreatingFlightId(),
+            fsObject.getCreatingFlightId())) {
+            throw new CorruptMetadataException("Unexpected flight id on file");
+        }
+
+        if (file.getObjectType() == FSObject.FSObjectType.FILE_NOT_PRESENT) {
+            return;
+        }
+
+        String sql = "UPDATE fs_object SET object_type = :object_type WHERE object_id = :object_id";
+        MapSqlParameterSource params = new MapSqlParameterSource()
+            .addValue("object_id", fsObject.getObjectId())
+            .addValue("object_type", FSObject.FSObjectType.FILE_NOT_PRESENT.toLetter());
+        jdbcTemplate.update(sql, params);
+    }
+
+    public boolean deleteFile(UUID objectId) {
+        return deleteFileWorker(objectId, null);
+    }
+
+    public boolean deleteFileForUndo(UUID objectId, String flightId) {
+        return deleteFileWorker(objectId, flightId);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    private boolean deleteFileWorker(UUID objectId, String flightId) {
```

Comment:
jeremy and i talked about the transaction isolation levels and for this, it should be read repeatable instead of the default read committed.

Reply:
I added DR-333 for implementing the change of isolation level and the retry for transaction conflict for this case.